### PR TITLE
Add support for entity discovery

### PR DESCRIPTION
**Pull request by eculver**
This adds a new `registry` package which implements support for "auto-discovery" of DOSA entities. It was broken out of #151. This will be used in subsequent changes to add support for UberFx.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -21,15 +21,51 @@
...
 type Config struct {
-	Scope      string       `yaml:"scope"`
-	NamePrefix string       `yaml:"namePrefix"`
-	Yarpc      yarpc.Config `yaml:"yarpc"`
+	Scope       string         `yaml:"scope"`
+	NamePrefix  string         `yaml:"namePrefix"`
+	Connector   string         `yaml:"connector"`
+	EntityPaths []string       `yaml:"entityPaths"`
+	Excludes    []string       `yaml:"excludes"`
+	Yarpc       yarpc.Config   `yaml:"yarpc"`
```
Comment (klnusbaum) [inline config/config.go]: This makes me a little nervous. Why is connector specific configuration being placed in the main configuration for DOSA? It's like the Yarpc connector is special for some reason.

File: config/config.go [hunk 2]
```diff
@@ -21,15 +21,51 @@
...
 
 import (
+	"time"
+
 	"github.com/uber-go/dosa"
 	"github.com/uber-go/dosa/connectors/yarpc"
 )
 
+const (
+	_defaultConn = "yarpc"
```
Comment (eculver) [inline config/config.go]: 👍

File: config/config.go (hunk 1 above)
Comment (eculver) [inline config/config.go]: It is is "special" in that it requires configuration whereas all other connectors do not require configuration. The config itself is in the `yarpc` package.

Also, if we remove it, we'll break backwards compatibility.

:shrug:

Comment (klnusbaum) [inline config/config.go]: sad times

Comment (klnusbaum) [inline config/config.go]: we should deprecate this field then. And the Yarpc connector should be able to just configure itself. That's beyond the scope of this PR though. When you merge this, I'll get to work on it.

Comment (rkuris) [inline config/config.go]: IMHO the right way to handle this is to make the configuration a set of K/V pairs. The connector supports having arbitrary K/V pairs.

A second possible implementation is to just pass the yaml parts down to the connector. The attribute map has {}interface as the value anyway, so you can just not parse it and pass it downstream.

Comment (klnusbaum) [inline config/config.go]: Yea, I think the second bit is the way to go.

**Pull request merged**